### ISSUE TITLE
Ruby: Seg fault in add_serialized_file with v3.18.0+

### ISSUE THREAD

**Issue by stanhu** [protocolbuffers/protobuf#9141]
**What version of protobuf and what language are you using?**
Version: master, v3.18.0, v3.19.0
Language: Ruby

**What operating system (Linux, Windows, ...) and version?**

s390x Linux on Z

**What runtime / compiler are you using (e.g., python version or gcc version)**

```
$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**What did you do?**
1. Use a IBM s390x machine running Linux on Z.
2. Clone the repo.
3. In the `ruby` dir, run `bundle exec rake`.

**What did you expect to see**

All tests successful

**What did you see instead?**

This seg fault:

```
/build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7: [BUG] Segmentation fault at 0x00000403ade27000
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [s390x-linux]

-- Control frame information -----------------------------------------------
c:0015 p:---- s:0062 e:000061 CFUNC  :add_serialized_file
c:0014 p:0053 s:0057 e:000056 TOP    /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7 [FINISH]
c:0013 p:---- s:0053 e:000052 CFUNC  :require
c:0012 p:0005 s:0048 e:000047 TOP    /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5 [FINISH]
c:0011 p:---- s:0045 e:000044 CFUNC  :require
c:0010 p:0083 s:0040 e:000039 TOP    /build/protobuf/ruby/lib/google/protobuf.rb:56 [FINISH]
c:0009 p:---- s:0037 e:000036 CFUNC  :require
c:0008 p:0005 s:0032 e:000031 TOP    /build/protobuf/ruby/tests/basic_test_pb.rb:4 [FINISH]
c:0007 p:---- s:0029 e:000028 CFUNC  :require
c:0006 p:0034 s:0024 e:000023 TOP    /build/protobuf/ruby/tests/basic.rb:6 [FINISH]
c:0005 p:---- s:0021 e:000020 CFUNC  :require
c:0004 p:0076 s:0016 e:000015 BLOCK  /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21 [FINISH]
c:0003 p:---- s:0011 e:000010 CFUNC  :select
c:0002 p:0018 s:0007 E:000cb8 EVAL   /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6 [FINISH]
c:0001 p:0000 s:0003 E:000750 (none) [FINISH]
```

[ruby-protobuf.txt](https://github.com/protocolbuffers/protobuf/files/7401074/ruby-protobuf.txt)

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

This is the log when I run `valgrind ruby tests/basic.rb`:

```
==140403== Invalid read of size 4
==140403==    at 0x9F97614: upb_fielddef_type (ruby-upb.c:4748)
==140403==    by 0x9F97614: upb_fielddef_isstring (ruby-upb.c:4923)
==140403==    by 0x9F97614: upb_fielddef_isprimitive (ruby-upb.c:4932)
==140403==    by 0x9F97614: create_fielddef (ruby-upb.c:6120)
==140403==    by 0x9F9771B: create_fielddef (ruby-upb.c:6029)
==140403==  Address 0x409fa3a2c is not stack'd, malloc'd or (recently) free'd
```

Full `valgrind` log:

[valgrind-protobuf.txt](https://github.com/protocolbuffers/protobuf/files/7401077/valgrind-protobuf.txt)

**Anything else we should know about your project / environment**

This is a Big Endian machine, and I checked that https://github.com/protocolbuffers/upb/pull/306/files seems to be in the `ruby-ubp.c` file. 

I didn't see these seg faults when I ran this test with v3.17.3. It seems to have been introduced in v3.18.0.

/cc: @haberman

**Comment by stanhu** [comment]
Using`git bisect`, I believe this seg fault started with 16e16eb8cda3035a93e4b3797336d351b07519da.

**Comment by stanhu** [comment]
I noticed that the failure was happening pretty early on for memory that appeared to be properly allocated, and the seg faults were happening perhaps due to some invalid memory address of a function. I noticed these warnings with `-O3` enabled:

```
../../../../ext/google/protobuf_c/ruby-upb.c: In function '_upb_decode':
../../../../ext/google/protobuf_c/ruby-upb.c:949:30: warning: argument 'buf' might be clobbered by 'longjmp' or 'vfork' [-Wclobbered]
 bool _upb_decode(const char *buf, size_t size, void *msg,
                  ~~~~~~~~~~~~^~~
../../../../ext/google/protobuf_c/ruby-upb.c: In function '_upb_decode.constprop':
../../../../ext/google/protobuf_c/ruby-upb.c:949:30: warning: argument 'buf' might be clobbered by 'longjmp' or 'vfork' [-Wclobbered]
../../../../ext/google/protobuf_c/ruby-upb.c: At top level:
```

I wasn't able to make the warnings go away by adding a `volatile` to `buf` (maybe that's the wrong place for it?), but when I compile with `-O0`, it seems that I was able to make more progress since the seg faults go away. Now, I'm running into validation issues with this serialized `descriptor_pb.rb`, which I've attached:

```
Traceback (most recent call last):
        13: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
        12: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
        11: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
        10: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `require'
         9: from /build/protobuf/ruby/tests/basic.rb:7:in `<top (required)>'
         8: from /build/protobuf/ruby/tests/basic.rb:7:in `require'
         7: from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in `<top (required)>'
         6: from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in `require'
         5: from /build/protobuf/ruby/lib/google/protobuf.rb:56:in `<top (required)>'
         4: from /build/protobuf/ruby/lib/google/protobuf.rb:56:in `require'
         3: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in `<top (required)>'
         2: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in `require'
         1: from /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7:in `<top (required)>'
/build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7:in `add_serialized_file': Unable to build file to DescriptorPool: Invalid default 'false' for field google.protobuf.EnumOptions.deprecated (Google::Protobuf::TypeError)
```

[descriptor_pb.rb.txt](https://github.com/protocolbuffers/protobuf/files/7448240/descriptor_pb.rb.txt)

Are these "invalid default" messages just missing implementations for these types? I can make more progress by converting the hard failure to a warning:

```
Invalid default 'false' for field google.protobuf.MethodDescriptorProto.client_streaming
Invalid default 'false' for field google.protobuf.MethodDescriptorProto.server_streaming
Invalid default 'false' for field google.protobuf.FileOptions.java_multiple_files
Invalid default 'false' for field google.protobuf.FileOptions.java_string_check_utf8
Invalid default 'SPEED' for field google.protobuf.FileOptions.optimize_for
Invalid default 'false' for field google.protobuf.FileOptions.cc_generic_services
Invalid default 'false' for field google.protobuf.FileOptions.java_generic_services
Invalid default 'false' for field google.protobuf.FileOptions.py_generic_services
Invalid default 'false' for field google.protobuf.FileOptions.php_generic_services
Invalid default 'false' for field google.protobuf.FileOptions.deprecated
Invalid default 'true' for field google.protobuf.FileOptions.cc_enable_arenas
Invalid default 'false' for field google.protobuf.MessageOptions.message_set_wire_format
Invalid default 'false' for field google.protobuf.MessageOptions.no_standard_descriptor_accessor
Invalid default 'false' for field google.protobuf.MessageOptions.deprecated
Invalid default 'STRING' for field google.protobuf.FieldOptions.ctype
Invalid default 'JS_NORMAL' for field google.protobuf.FieldOptions.jstype
Invalid default 'false' for field google.protobuf.FieldOptions.lazy
Invalid default 'false' for field google.protobuf.FieldOptions.deprecated
Invalid default 'false' for field google.protobuf.FieldOptions.weak
Invalid default 'false' for field google.protobuf.EnumOptions.deprecated
Invalid default 'false' for field google.protobuf.EnumValueOptions.deprecated
Invalid default 'false' for field google.protobuf.ServiceOptions.deprecated
Invalid default 'false' for field google.protobuf.MethodOptions.deprecated
Invalid default 'IDEMPOTENCY_UNKNOWN' for field google.protobuf.MethodOptions.idempotency_level
```

But then I run into this error:

```
Traceback (most recent call last):
        16: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
        15: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
        14: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
        13: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `require'
        12: from /build/protobuf/ruby/tests/basic.rb:6:in `<top (required)>'
        11: from /build/protobuf/ruby/tests/basic.rb:6:in `require'
        10: from /build/protobuf/ruby/tests/basic_test_pb.rb:6:in `<top (required)>'
         9: from /build/protobuf/ruby/tests/basic_test_pb.rb:6:in `require'
         8: from /build/protobuf/ruby/tests/google/protobuf/wrappers_pb.rb:6:in `<top (required)>'
         7: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:453:in `build'
         6: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:453:in `instance_eval'
         5: from /build/protobuf/ruby/tests/google/protobuf/wrappers_pb.rb:7:in `block in <top (required)>'
         4: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:34:in `add_file'
         3: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:34:in `new'                                                              2: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:71:in `initialize'
         1: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:71:in `new'
/build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:71:in `initialize': Expected number type for integral field 'name' (given String)
. (Google::Protobuf::TypeError)
```

I'm not sure why the `name` field is expecting a number, but it sounds like there was some problem defining the types.

I noticed on my macOS system the serialized format descriptor wasn't present in `descriptor_pb.rb`, so I'm wondering if I'm just running into a code path not used often?

**Comment by stanhu** [comment]
It seems that for `descriptor.proto` those invalid defaults are coming from boolean values, which is represented by `UPB_DESCRIPTOR_TYPE_BOOL` (8): https://github.com/protocolbuffers/protobuf/blob/26603a14666c0f30883bc4508591522586ba51f8/ruby/ext/google/protobuf_c/ruby-upb.h#L542

However, the `parse_default` function is interpreting this value as `UPB_TYPE_INT64` in https://github.com/protocolbuffers/protobuf/blob/26603a14666c0f30883bc4508591522586ba51f8/ruby/ext/google/protobuf_c/ruby-upb.c#L5874?

This might also explain why the `name` field appears to have the wrong type as well.

Does `descriptor.proto` have to be treated differently here?

**Comment by stanhu** [comment]
Never mind, somehow `f->type_` is actually 0. Somehow that's getting mapped to 8 when I do the following:

```diff
@@ -5933,8 +5937,8 @@ static void parse_default(symtab_addctx *ctx, const char *str, size_t len,
   return;

 invalid:
-  symtab_errf(ctx, "Invalid default '%.*s' for field %s", (int)len, str,
-              upb_fielddef_fullname(f));
+  fprintf(stderr, "Invalid default '%.*s' for field %s, type %d\n", (int)len, str,
+         upb_fielddef_fullname(f), upb_fielddef_type(f));
 }
```

But if I add this diff:

```diff
diff --git a/ruby/ext/google/protobuf_c/ruby-upb.c b/ruby/ext/google/protobuf_c/ruby-upb.c
index d68caac0e..0d0444a40 100755
--- a/ruby/ext/google/protobuf_c/ruby-upb.c
+++ b/ruby/ext/google/protobuf_c/ruby-upb.c
@@ -4776,6 +4776,7 @@ upb_fieldtype_t upb_fielddef_type(const upb_fielddef *f) {
     case UPB_DESCRIPTOR_TYPE_MESSAGE:
       return UPB_TYPE_MESSAGE;
   }
+  fprintf(stderr, "upb_fielddef_type() unreachable!\n");
   UPB_UNREACHABLE();
 }
```

Somehow the assert gets triggered, since the value is `type_` is indeed 0?

**Comment by haberman** [comment]
Hi @stanhu, thanks for all of the debugging info and sorry for the slow reply.

The Ruby library is built on https://github.com/protocolbuffers/upb, which has undergone a number of changes that have not yet propagated to Ruby. So my first question is whether the problem might possibly have been fixed in upb HEAD already.

I am wondering if you get any errors when running the following on s390x:

```
$ git clone https://github.com/protocolbuffers/upb.git
$ cd upb
$ bazel test ...
```

Also would it be possible to get access to an s390x machine to debug on?

**Comment by haberman** [comment]
> Also would it be possible to get access to an s390x machine to debug on?

Ah nevermind, I just saw your link above to a list of services for testing on s390x.

**Comment by stanhu** [comment]
> I am wondering if you get any errors when running the following on s390x

I actually tried to do this, but as far as I know `bazel` isn't available on s390x (https://github.com/bazelbuild/bazel/issues/10408). I was able to build the library and compile some of the tests manually, but I wasn't able to induce any errors with at least some of the tests. So perhaps we do need to pull in these changes.

> Also would it be possible to get access to an s390x machine to debug on?

I will ask nonetheless!

**Comment by stanhu** [comment]
@haberman Is there a script to generate the `ruby-upb.{c,h}` files from the upb directory?

**Comment by haberman** [comment]
With upb checked out, run:

```
$ bazel build :gen_ruby_amalgamation
```

Then you can copy that into the protobuf repo.

Since there is a bit of work involved in doing this update (some upb interfaces have had breaking changes) I did the update in this branch for you to try out: https://github.com/haberman/protobuf/tree/ruby-upb

Do you want to give that branch a try and see if it still crashes on s390x?

**Comment by stanhu** [comment]
@haberman Thanks! I tried out the branch, but I seem to be missing `utf8_range`:

```
In file included from ../../../../ext/google/protobuf_c/protobuf.h:38,
                 from ../../../../ext/google/protobuf_c/protobuf.c:31:
../../../../ext/google/protobuf_c/ruby-upb.h:760:10: fatal error: third_party/utf8_range/utf8_range.h: No such file or directory
 #include "third_party/utf8_range/utf8_range.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by haberman** [comment]
@stanhu did you do `rake test`? I set up the Rakefile to copy those files from `third_party/utf8_range` in the base directory of the repo.

**Comment by stanhu** [comment]
I thought I did do `rake test` and `rake compile`, but I didn't see it copied. I manually copied things over. Now I get:

```
cp tests/basic.rb tmp/s390x-linux/stage/tests/basic.rb
cp tests/generated_code_test.rb tmp/s390x-linux/stage/tests/generated_code_test.rb
cp tests/stress.rb tmp/s390x-linux/stage/tests/stress.rb
install -c tmp/s390x-linux/protobuf_c/2.7.2/protobuf_c.so lib/google/protobuf_c.so
cp tmp/s390x-linux/protobuf_c/2.7.2/protobuf_c.so tmp/s390x-linux/stage/lib/google/protobuf_c.so
/build/protobuf/ruby/tests/basic.rb:140: warning: ambiguous first argument; put parentheses or a space even after `-' operator
/build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:4: warning: /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:4: warning: loading in progress, circular require considered harmful - /build/protobuf/ruby/lib/google/protobuf.rb
        from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in  `<main>'
        from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in  `select'
        from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in  `block in <main>'
        from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in  `require'
        from /build/protobuf/ruby/tests/basic.rb:6:in  `<top (required)>'
        from /build/protobuf/ruby/tests/basic.rb:6:in  `require'
        from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in  `<top (required)>'
        from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in  `require'
        from /build/protobuf/ruby/lib/google/protobuf.rb:56:in  `<top (required)>'
        from /build/protobuf/ruby/lib/google/protobuf.rb:56:in  `require'
        from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in  `<top (required)>'
        from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in  `require'
        from /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:4:in  `<top (required)>'
        from /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:4:in  `require'

Traceback (most recent call last):
        13: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
        12: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
        11: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
        10: from /usr/local/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `require'
         9: from /build/protobuf/ruby/tests/basic.rb:6:in `<top (required)>'
         8: from /build/protobuf/ruby/tests/basic.rb:6:in `require'
         7: from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in `<top (required)>'
         6: from /build/protobuf/ruby/tests/basic_test_pb.rb:4:in `require'
         5: from /build/protobuf/ruby/lib/google/protobuf.rb:56:in `<top (required)>'
         4: from /build/protobuf/ruby/lib/google/protobuf.rb:56:in `require'
         3: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in `<top (required)>'
         2: from /build/protobuf/ruby/lib/google/protobuf/descriptor_dsl.rb:5:in `require'
         1: from /build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7:in `<top (required)>'
/build/protobuf/ruby/lib/google/protobuf/descriptor_pb.rb:7:in `add_serialized_file': Unable to build file to DescriptorPool: invalid type for field google.protobuf.FileDescriptorSet.file (0) (Google::Protobuf::TypeError)
rake aborted!
```

**Comment by stanhu** [comment]
Attached is the full output from `rake test`. Notice the `longjmp` warnings are still there:

[rake.txt](https://github.com/protocolbuffers/protobuf/files/7457728/rake.txt)

I didn't see any difference this time with `-O0`, but I think it's failing earlier than before.

**Comment by stanhu** [comment]
I got `bazel` to compile thanks to https://github.com/bazelbuild/bazel/issues/13597#issuecomment-930606195.

```
TIMEOUT: //tests:test_conformance_upb (Summary)
      /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/tests/test_conformance_upb/test.log
INFO: Elapsed time: 721.923s, Critical Path: 318.87s
INFO: 3406 processes: 416 internal, 1 local, 2989 processwrapper-sandbox.
INFO: Build completed, 6 tests FAILED, 3406 total actions
//cmake:cmake_build                                                      PASSED in 3.9s
//cmake:test_generated_files                                             PASSED in 0.3s
//python:minimal_test                                                    PASSED in 0.1s
//tests:file_descriptor_parsenew_fuzzer                                  PASSED in 0.1s
//tests:test_generated_code                                              PASSED in 0.1s
//tests:test_table                                                       PASSED in 2.3s
//tests:test_conformance_upb                                            TIMEOUT in 300.1s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/tests/test_conformance_upb/test.log
//:msg_test                                                              FAILED in 0.2s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/msg_test/test.log
//tests:test_cpp                                                         FAILED in 0.2s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/tests/test_cpp/test.log
//tests/bindings/lua:test_lua                                            FAILED in 0.2s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/tests/bindings/lua/test_lua/test.log
//upb/util:def_to_proto_test                                             FAILED in 0.4s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/upb/util/def_to_proto_test/test.log
//upb/util:required_fields_test                                          FAILED in 1.3s
  /root/.cache/bazel/_bazel_root/6bf3da2484ef4bca9a0fbec97f98a05c/execroot/upb/bazel-out/s390x-fastbuild/testlogs/upb/util/required_fields_test/test.log

Executed 12 out of 12 tests: 6 tests pass and 6 fail locally.
INFO: Build completed, 6 tests FAILED, 3406 total actions
```

[testlogs.tar.gz](https://github.com/protocolbuffers/protobuf/files/7458901/testlogs.tar.gz)

This log looks particularly interesting:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tests:test_conformance_upb
-----------------------------------------------------------------------------
Error loading compiled-in descriptor for file 'src/google/protobuf/test_messages_proto2.proto' (this should never happen): invalid type for field protobuf_test_messages.proto2.TestAllTypesProto2.optional_nested_message (0)
Error loading compiled-in descriptor for file 'google/protobuf/struct.proto' (this should never happen): invalid type for field google.protobuf.Struct.fields (0)
Error loading compiled-in descriptor for file 'src/google/protobuf/test_messages_proto3.proto' (this should never happen):
-- Test timed out at 2021-11-02 08:04:15 UTC --
```

**Comment by stanhu** [comment]
Seems related to https://github.com/protocolbuffers/upb/pull/435.

**Comment by stanhu** [comment]
I've confirmed that https://github.com/protocolbuffers/upb/pull/435 eliminates the seg fault when I apply it the protobuf master. I do see this other failure:

```
Failure: test_defined_defaults(BasicTestProto2::MessageContainerTest)
/build/protobuf/ruby/tests/basic_proto2.rb:118:in `test_defined_defaults'
     115:
     116:     def test_defined_defaults
     117:       m = TestMessageDefaults.new
  => 118:       assert_equal 1, m.optional_int32
     119:       assert_equal 2, m.optional_int64
     120:       assert_equal 3, m.optional_uint32
     121:       assert_equal 4, m.optional_uint64
<1> expected but was
<0>
===========================================================================================================================================================================================================================
.........................................../build/protobuf/ruby/tests/common_tests.rb:355: warning: instance variable descriptor not initialized
../build/protobuf/ruby/tests/common_tests.rb:304: warning: instance variable descriptor not initialized
..F
===========================================================================================================================================================================================================================
Failure: test_set_clear_defaults(BasicTestProto2::MessageContainerTest)
/build/protobuf/ruby/tests/basic_proto2.rb:148:in `test_set_clear_defaults'
     145:       assert_equal -42, m.optional_int32
     146:       assert m.has_optional_int32?
     147:       m.clear_optional_int32
  => 148:       assert_equal 1, m.optional_int32
     149:       assert !m.has_optional_int32?
     150:
     151:       m.optional_string = "foo bar"
<1> expected but was
<0>
===========================================================================================================================================================================================================================
.............................................................../build/protobuf/ruby/tests/type_errors.rb:158: warning: instance variable descriptor not initialized
/build/protobuf/ruby/tests/type_errors.rb:162: warning: instance variable descriptor not initialized
..........
Finished in 2.494228085 seconds.
```

**Comment by stanhu** [comment]
Ok, all `protobuf` and `upb` tests now pass once I regenerated `ruby-upb.{c,h}` with `bazel gen_ruby_amalgamation` in the `upb` project with the following pull requests:

1. https://github.com/protocolbuffers/upb/pull/435
2. https://github.com/protocolbuffers/upb/pull/441
3. https://github.com/protocolbuffers/upb/pull/445
4. https://github.com/protocolbuffers/upb/pull/448
5. https://github.com/protocolbuffers/upb/pull/450
6. https://github.com/protocolbuffers/upb/pull/437
7. https://github.com/protocolbuffers/upb/pull/440

**Comment by stanhu** [comment]
@haberman What do we need to do to get https://github.com/haberman/protobuf/tree/ruby-upb (and the latest upb) ready for review?

**Comment by haberman** [comment]
All of the upb changes for big endian should have landed for Ruby by now. I'm going to close this bug.

**Comment by stanhu** [comment]
Thanks, to be clear this landed in v3.21.0 via https://github.com/protocolbuffers/protobuf/pull/9589.
